src/UsgsAstroPlugin.cpp

```diff
@@ -0,0 +1,306 @@
+  if (modelName != model_name_from_state ||
+      modelName != UsgsAstroFrameSensorModel::_SENSOR_MODEL_NAME){
+          constructible = false;
+      }
+  // Check that the necessary keys are there (this does not chek values at all.)
+  // This sort of check should probably be moved into the sensor model
+  // auto state = json::parse(modelState);
+  // for(auto &key : MODEL_KEYWORDS[model_name_from_state]){
+  //     if (state.find(key) == state.end()){
+  //         constructible = false;
+  //         break;
+  //     }
+  // }
+  return constructible;
+}
+
+
+bool UsgsAstroPlugin::canModelBeConstructedFromISD(const csm::Isd &imageSupportData,
+                                              const std::string &modelName,
+                                              csm::WarningList *warnings) const {
+  return canISDBeConvertedToModelState(imageSupportData, modelName, warnings);
+}
+
+
+// This function takes a csm::Isd which only has the image filename set. It uses this filename to
+// find a metadata json file loacated alongside the image file. It creates and returns new csm::Isd
+// with its parameters populated by the metadata file.
+std::string UsgsAstroPlugin::loadImageSupportData(const csm::Isd &imageSupportDataOriginal) const {
+
+  // Get image location from the input csm::Isd:
+  std::string imageFilename = imageSupportDataOriginal.filename();
+  size_t lastIndex = imageFilename.find_last_of(".");
+  std::string baseName = imageFilename.substr(0, lastIndex);
+  std::string isdFilename = baseName.append(".json");
+
+  try {
+    std::ifstream isd_sidecar(isdFilename);
+    json jsonisd;
+    isd_sidecar >> jsonisd;
+    return jsonisd.dump();
+
+  } catch (...) {
+    std::string errorMessage = "Could not read metadata file associated with image: ";
+    errorMessage.append(isdFilename);
+    throw csm::Error(csm::Error::FILE_READ, errorMessage,
+                     "UsgsAstroPlugin::loadImageSupportData");
+  }
+}
+
+
+std::string UsgsAstroPlugin::getModelNameFromModelState(const std::string &modelState,
+                                                   csm::WarningList *warnings) const {
+  auto state = json::parse(modelState);
+
+  std::string name = state.value<std::string>("model_name", "");
+
+  if (name == "") {
+      csm::Error::ErrorType aErrorType = csm::Error::INVALID_SENSOR_MODEL_STATE;
+      std::string aMessage = "No 'model_name' key in the model state object.";
+      std::string aFunction = "UsgsAstroPlugin::getModelNameFromModelState";
+      csm::Error csmErr(aErrorType, aMessage, aFunction);
+      throw(csmErr);
+  }
+
+  // if (MODEL_KEYWORDS.find(name) == MODEL_KEYWORDS.end()){
+  //     csm::Error::ErrorType aErrorType = csm::Error::SENSOR_MODEL_NOT_SUPPORTED;
+  //     std::string aMessage = "Sensor model not supported.";
+  //     std::string aFunction = "UsgsAstroPlugin::getModelNameFromModelState()";
+  //     csm::Error csmErr(aErrorType, aMessage, aFunction);
+  //     throw(csmErr);
+  // }
```

Comment:
Why's this here? Do we need differentiate between invalid state for supported model and unsupported model?

Reply:
Most of it was commented out while I was trying to figure out the best way to handle these errors. The original was mostly spaghetti code and leaving them as they were caused stack overflow errors, but at the same time wasn't sure what exactly the CSM spec wanted in terms of where errors should be handled and how those errors should look like. This is pretty much the story behind most of the commented out errors and validation checking.